### PR TITLE
Units in solutions

### LINKED ISSUE #541: solution array __call__ should return units
When a user does `sol("W")`, the output should have units. Otherwise, incorrect computations involving mismatched units could be done by an unsuspecting user.

MWE:
```python
from gpkit import Model, Variable

x = Variable("x", 10, "ft")
y = Variable("y", "m")
m = Model(y, [y >= x])
sol = m.solve(verbosity=0)
print sol("y")/sol("x")
```
(should print 1.0; prints `0.304799996606`)

### PR DESCRIPTION
**Pull request by bqpd**
Closes #541.

@whoburg, note that this implements Quantity substitution but does not test it outside of cvxopt cost functions.

### CODE REVIEW

**Comment by whoburg** [comment]
@galbramc We're getting a `Unable to import 'gpkit'` pylint warning in `docs/source/examples`. Any idea what we're doing wrong? Do the pylint machines install gpkit before running pylint on the examples dir?

**Comment by whoburg** [comment]
@bqpd can you explain what you mean by it "does not test it outside of cvxopt cost functions"?

**Comment by bqpd** [comment]
Oh, the only quantity substitutions that occur during tests are
substitutions into the cost, since cvxopt doesn't return the cost. We
should add more tests before saying Quantity substitution works.

**Comment by galbramc** [comment]
Add this to pylint.sh

```
# Add gpkit to the python path so that pylint can import gpkit when analyzing the examples directory
export PYTHONPATH=$PYTHONPATH:$WORKSPACE/gpkit/
```

That should take care of the import errors for pylint. I don't actually install gpkit during the testing, which is why you get the import error.

**Comment by whoburg** [comment]
retest this please

**Comment by whoburg** [comment]
retest this please

**Comment by galbramc** [comment]
Looks like the beam test has some unit problems. Only x is defined in meters, but EI, q, and L look like they are missing units and hence causing failures.

**Comment by whoburg** [comment]
@bqpd fyi I force pushed a rebase to clean up a long string of pylint commits. If you pulled from this branch in the last hour you may need to re-checkout the branch

yeah, we have a test failure on most testing machines:
```python
Traceback (most recent call last):
  File "/home1/jenkins/workspace/gpkit_PullRequest/buildnode/reynolds/optimizer/mosek/gpkit/tests/helpers.py", line 51, in test
    logged_example_testcase(name, import_dict, path)(self)
  File "/home1/jenkins/workspace/gpkit_PullRequest/buildnode/reynolds/optimizer/mosek/gpkit/tests/helpers.py", line 67, in test
    imported[name] = importlib.import_module(name)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home1/jenkins/workspace/gpkit_PullRequest/buildnode/reynolds/optimizer/mosek/docs/source/examples/beam.py", line 60, in <module>
    w_exact = q/(24.*EI) * x**2 * (x**2 - 4*L*x + 6*L**2)  # analytic soln
  File "/home1/jenkins/.local/lib/python2.7/site-packages/pint/quantity.py", line 504, in __sub__
    return self._add_sub(other, operator.sub)
  File "/home1/jenkins/.local/lib/python2.7/site-packages/pint/quantity.py", line 415, in _add_sub
    other.dimensionality)
DimensionalityError: Cannot convert from 'meter ** 4' ([length] ** 4) to 'meter ** 3' ([length] ** 3)
		
```
interestingly it does not fail on Macys or reynolds_windows7

**Comment by bqpd** [comment]
Yeah, that shouldn't fail. :p what could be OS dependent?

**Comment by galbramc** [comment]
I don't know. Pint is installed on macys, but looking at the output from the log it does not look like the tests ran twice (once with and once without units)

**Comment by whoburg** [comment]
also strange: unit tests pass on my computer on this branch both before and after merging to master

**Comment by whoburg** [comment]
oh the elcap error is different:

```python
Stacktrace

Traceback (most recent call last):
  File "/Users/jenkins/workspace/gpkit_PullRequest/buildnode/macys_ElCap/optimizer/cvxopt/gpkit/tests/helpers.py", line 51, in test
    logged_example_testcase(name, import_dict, path)(self)
  File "/Users/jenkins/workspace/gpkit_PullRequest/buildnode/macys_ElCap/optimizer/cvxopt/gpkit/tests/helpers.py", line 67, in test
    imported[name] = importlib.import_module(name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/jenkins/workspace/gpkit_PullRequest/buildnode/macys_ElCap/optimizer/cvxopt/docs/source/examples/beam.py", line 58, in <module>
    x = np.linspace(0, L, len(q))*gpkit.units.m  # position along beam
  File "/Library/Python/2.7/site-packages/numpy-1.10.1-py2.7-macosx-10.11-intel.egg/numpy/core/function_base.py", line 115, in linspace
    y[-1] = stop
ValueError: setting an array element with a sequence.
		
```

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
I upgraded pint on the testing machines, that seemed to fix some problems, but also created new ones

**Comment by whoburg** [comment]
@bqpd, did a quick code review for your consideration in addition to dealing with the failing tests

**Comment by whoburg** [comment]
whoa, @galbramc, looking at those errors I suspect we'll fail the next time someone pushes to master. Thanks for catching the fact that pint changed... but we might need to address that in a separate issue in case it takes time.

@bqpd, any thoughts? should Marshall revert pint for now?

**Comment by bqpd** [comment]
The elcap error makes sense, that seems like a numpy difference (but should be fixed in GPkit). @whoburg see what I mean about how the inability to set an array element to a quantity just propagates?

**Comment by whoburg** [comment]
is our fundamental problem with using Quantity(numpy_array, units) objects (as described in the pint docs) that we want each element of the array to be allowed to have differing units?

**Comment by bqpd** [comment]
Not at all. It's that I want to be able to have an array with quantities as objects. Even though all the quantities have the same units, I want to say `a[0] = q`. If I could do this, PosyArrays could contain quantities, etc.

**Comment by whoburg** [comment]
Confused about where it fails. See below

```python
In [1]: import numpy as np

In [2]: from pint import UnitRegistry

In [3]: Quantity = UnitRegistry().Quantity

In [4]: np.array([Quantity(3, "m"), Quantity(42, "fortnights"), Quantity(3, "ft")], dtype="object")
Out[4]:
array([<Quantity(3, 'meter')>, <Quantity(42, 'fortnight')>,
       <Quantity(3, 'foot')>], dtype=object)

In [5]: a = _

In [6]: a[1] = Quantity(3, "lbf")

In [7]: a
Out[7]:
array([<Quantity(3, 'meter')>, <Quantity(3, 'force_pound')>,
       <Quantity(3, 'foot')>], dtype=object)
```

**Comment by bqpd** [comment]
huh! Will get a MWE of the "cannot set array element with sequence" error that I always see.

**Comment by whoburg** [comment]
I get that without `dtype="object"` -- is that the issue?

**Comment by bqpd** [comment]
Hmm...

**Comment by galbramc** [comment]
test this please

**Comment by galbramc** [comment]
Here are the different numpy versions on the three machines

reynolds: 1.8.2
macys: 1.9.2
macys-ElCap: 1.10.1

Could that be the cause of mixed failures/success?

**Comment by galbramc** [comment]
Btw, I did revert pint back to version 0.6.

It might be a good idea to print the version number of your dependencies as part of the testing script. That way when someone else has problems you can ask for the complete output which will include versions. Just a thought.

**Comment by galbramc** [comment]
I would put fixing the pint 0.7 issues pretty high on the list. If as a user, I follow your install inductions,   gpkit will not work after I install pint and get the latest version. So passing on the testing machines right now is somewhat of a false positive.

**Comment by bqpd** [comment]
retest this please

**Comment by bqpd** [comment]
@whoburg, this PR is separate from the pints issues; shall we merge it if it passes tests?

**Comment by galbramc** [comment]
test this please

**Comment by bqpd** [comment]
test this please

**Comment by bqpd** [comment]
Now pasing tests, woot.

**Comment by bqpd** [comment]
retest this please (@galbramc, any idea why dependency is failing here again?)

**Comment by bqpd** [comment]
retest this please

**Comment by galbramc** [comment]
Apparently the file path at the top of the of the file that specifies the interpreter was too long. So I had to change the script to explicitly use python to execute the pip script.

**Comment by bqpd** [comment]
bizarre!

**Comment by bqpd** [comment]
@whoburg, after renaming `enray_dict` I think this is ready to merge per your previous code review.

**Comment by bqpd** [comment]
Pretty sure linspace in some versions of numpy might assume floats...

**Comment by galbramc** [comment]
test this please

File: gpkit/small_classes.py [hunk 1]
```diff
@@ -137,16 +136,21 @@ def index_dict(idx, i, o):
...
-            else:
-                o[k] = np.array(v)
+                v = v[0]
+            v = np.array(v)
+            if (united and hasattr(k, "units")
+                    and isinstance(k.units, Quantity)):
+                v = v*k.units
+            o[k] = v
     # assert set(i.keys()) == set(o.keys())  # keys change with swept varkeys
     return o
```
Comment (whoburg) [inline gpkit/small_classes.py]: I see, it's really a help method for `DictOfLists`. Should it be private (i.e. `_enray_and_unit_dict`)?

Comment (bqpd) [inline gpkit/small_classes.py]: Seems reasonable!

Comment (bqpd) [inline gpkit/small_classes.py]: Done.

**Comment by whoburg** [comment]
@bqpd, regarding the pylint errors, I just went through and changed `i` to `d_in` and `o` to `d_out`. Should I go ahead and push that commit, or do you prefer to change pylint to allow the argument name `o`?

**Comment by bqpd** [comment]
Go ahead! I think it makes it slightly harder to read, but pretty marginal.

**Comment by whoburg** [comment]
planning to merge once tests pass

**Comment by bqpd** [comment]
(they passed)

**Pull request merged**